clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,341 @@
+ * Meter type is mostly a Total of a Count metric.
+ * We can assume that a Total metric always increases (but it is not guaranteed as the sample values might be both
+ * negative or positive).
+ * For now, Total is converted to CUMULATIVE_DOUBLE unless we find a valid counter-example.
+ *
+ * <p>
+ *
+ * The Sum as it is a sample sum which is not a cumulative metric. It is converted to GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * The compound metrics are virtual metrics. They are composed of simple types or anonymous measurable types
+ * which are reported. A compound metric is never reported as-is.
+ *
+ * <p>
+ *
+ * A Meter metric is always created with and reported as 2 KafkaExporter metrics: a rate and a
+ * count. For eg: org.apache.kafka.common.network.Selector has Meter metric for "connection-close" but it
+ * has to be created with a "connection-close-rate" metric of type rate and a "connection-close-total"
+ * metric of type total. So, we will never get a KafkaExporter metric with type Meter.
+ *
+ * <p>
+ *
+ * Frequencies is created with a array of Frequency objects. When a Frequencies metric is registered, each
+ * member Frequency object is converted into an anonymous Measurable and registered. So, a Frequencies metric
+ * is reported with a set of measurables with name = Frequency.name(). As there is no way to figure out the
+ * compound type, each component measurables is converted to a GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * Percentiles work the same way as Frequencies. The only difference is that it is composed of Percentile
+ * types instead. So, we should treat the component measurable as GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * Some metrics are defined as either anonymous inner classes or lambdas implementing the Measurable
+ * interface. As we do not have any information on how to treat them, we should fallback to treating
+ * them as GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * KafkaExporter -> OpenTelemetry mapping for measurables
+ * Avg / Rate / Min / Max / Total / Sum -> Gauge
+ * Count -> Sum
+ * Meter has 2 elements :
+ *  Total -> Sum
+ *  Rate -> Gauge
+ * Frequencies -> each component is Gauge
+ * Percentiles -> each component is Gauge
+ */
+public class KafkaMetricsCollector implements MetricsCollector {
+
+    private static final Logger log = LoggerFactory.getLogger(KafkaMetricsCollector.class);
+
+    private final StateLedger ledger;
+    private final Clock clock;
+    private final MetricNamingStrategy<MetricName> metricNamingStrategy;
+
+    private static final Field METRIC_VALUE_PROVIDER_FIELD;
+
+    static {
+        try {
+            METRIC_VALUE_PROVIDER_FIELD = KafkaMetric.class.getDeclaredField("metricValueProvider");
+            METRIC_VALUE_PROVIDER_FIELD.setAccessible(true);
+        } catch (Exception e) {
+            throw new KafkaException(e);
+        }
+    }
+
+    public KafkaMetricsCollector(MetricNamingStrategy<MetricName> metricNamingStrategy) {
+        this(metricNamingStrategy, Clock.systemUTC());
```

Comment:
Do you actually need to use a clock? Can we just use the time util?

Reply:
The code came from a repo which did not use Kafka internal classes, which is why it uses the Java standard library Clock, since Time may not be strictly part of the public AK APIs.

Reply:
Time/Timer is generally used, at least, in the client library.  It also has a bunch of mock wrappers around it, so it can be pretty handy when it comes to testing.

Reply:
Move to `Time`, done.